File: wpf/src/Circles/World.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;
using System.Diagnostics;


namespace Circles
{
    class World
    {
        public List<MovingCircle> Objects;
        public double Width { get; }
        public double Height { get; }
        private double Gravity;
        private Vector[] PrecalculatedDeltaFCache = null;

        public World(double width, double height, double Gravity) {
            this.Objects = new List<MovingCircle>();
            this.Width = width;
            this.Height = height;
            this.Gravity = Gravity;
        }

        public void Tick(double timestep, double gravity)
        {
            this.Gravity = gravity;
            Tick(timestep);
        }

        public void Tick(double timestep) {
            timestep /= 1000; // Just seconds, please

            int len = Objects.Count;
            Vector[] accelerationCache = new Vector[len];
            // We know that the vector value is 0, 0
            this.PrecalculatedDeltaFCache = new Vector[len];

            for (int i = 0; i < len; ++i) {
                MovingCircle obj = Objects[i];
                Vector acceleration = force(obj, i, len) / obj.Mass;

                obj.Position += timestep * (obj.Speed + timestep * acceleration * .5);

                accelerationCache[i] = acceleration;

                CalculateSpeedAndPositionFromCollisions(obj, i, len);
            }

            this.PrecalculatedDeltaFCache = new Vector[len];
            // We want updated position and speeds
            for (int i = 0; i < len; ++i) {
                MovingCircle obj = Objects[i];
                Vector newAcceleration = force(obj, i, len) / obj.Mass;
                // (oldAcceleration + newAcceleration) / 2
                obj.Speed += timestep * (accelerationCache[i] + newAcceleration) * .5;
            }
        }

        public void CalculateSpeedAndPositionFromCollisions(MovingCircle obj, int myIndex, int len) {
            int len = Objects.Count;

            // We only detect collisions for bodies after us in the list
            for (int i = myIndex + 1; i < len; ++i)
            {
                MovingCircle other = Objects[i];

                Vector centerToCenter = (other.Center - obj.Center);
                double centerToCenterLen = centerToCenter.Length;

                // If they collide
                if ( centerToCenterLen <= other.Radius + obj.Radius )
                {
                    // We get the translation necessary to move one
                    Vector a = centerToCenter * (obj.Radius / centerToCenterLen);
                    Vector b = centerToCenter * (centerToCenterLen - other.Radius) / centerToCenterLen;
                    Vector middle = (a - b) / 2;

                    obj.Position -= middle;
                    other.Position += middle;

                    // Now we've positioned the cicles, we can normalize the centerToCenter vector in order to convert it into a unit vector
                    centerToCenter.Normalize();

                    // http://ericleong.me/research/circle-circle/#static-circle-circle-collision-detection
                    double p = 2 * (obj.Speed.X * centerToCenter.X + obj.Speed.Y * centerToCenter.Y - other.Speed.X * centerToCenter.X - other.Speed.Y * centerToCenter.Y) / (obj.Mass + other.Mass);
                    obj.Speed = new Vector(obj.Speed.X - p * other.Mass * centerToCenter.X,
                                           obj.Speed.Y - p * other.Mass * centerToCenter.Y)
                                           * obj.BounceFactor;
                    other.Speed = new Vector(other.Speed.X + p * obj.Mass * centerToCenter.X,
                                             other.Speed.Y + p * obj.Mass * centerToCenter.Y)
                                             * other.BounceFactor;
                    /* Original: Seems to have switched both masses, since a high-mass sphere bounces like heck
                    obj.Speed = new Vector(obj.Speed.X - p * obj.Mass * centerToCenter.X,
                                           obj.Speed.Y - p * obj.Mass * centerToCenter.Y)
                                           * obj.BounceFactor;
                    other.Speed = new Vector(other.Speed.X + p * other.Mass * centerToCenter.X,
                                             other.Speed.Y + p * other.Mass * centerToCenter.Y)
                                             * other.BounceFactor;
                    */
                }
            }

            if (obj.Position.Y < 0) {
                obj.Speed = new Vector(obj.Speed.X, Math.Abs(obj.Speed.Y) * obj.BounceFactor);
                obj.Position = new Point(obj.Position.X, 0);
            } else if (obj.Position.Y + obj.Diameter > this.Height) {
                obj.Speed = new Vector(obj.Speed.X, -Math.Abs(obj.Speed.Y) * obj.BounceFactor);
                obj.Position = new Point(obj.Position.X, this.Height - obj.Diameter);
            }

            if (obj.Position.X < 0) {
                obj.Position = new Point(0, obj.Position.Y);
                obj.Speed = new Vector(-obj.Speed.X * obj.BounceFactor, obj.Speed.Y);
            } else if (obj.Position.X + obj.Diameter > this.Width) {
                obj.Position = new Point(this.Height - obj.Diameter, obj.Position.Y);
                obj.Speed = new Vector(-obj.Speed.X * obj.BounceFactor, obj.Speed.Y);
            }

        }

        public Vector force(MovingCircle obj, int myIndex, int len) {
            Vector f = PrecalculatedDeltaFCache[myIndex];

            // Vector f = new Vector(0.0, 9.8) * obj.Mass
            // Actually the gravity is in px/s/s... Lol
            f += new Vector(0.0, this.Gravity) * obj.Mass; // Gravity

            // We only detect collisions for bodies after us in the list
            for (int i = myIndex + 1; i < len; ++i)
            {
                MovingCircle other = Objects[i];

                Vector centerToCenter = (other.Center - obj.Center);

                // If they collide
                // NOTE: Instead of length < o1.radius + o2.radius we use LengthSquared, since it's faster
                // (Obtaining the length needs a square root)
                if ( centerToCenter.LengthSquared < other.Radius * other.Radius + obj.Radius * obj.Radius )
                {
                    // Now we've positioned the cicles, we can normalize the centerToCenter vector in order to convert it into a unit vector
                    centerToCenter.Normalize();

                    // Add the linear momentum and substract the normal
                    f += other.Speed * other.Mass;
                    f -= obj.Speed * obj.Mass;

                    PrecalculatedDeltaFCache[i] += obj.Speed * obj.Mass;
                    PrecalculatedDeltaFCache[i] -= other.Speed * other.Mass;
                }
            }

            // if ( object touches boundary )
            //     f += normal; (f = 0);
            // TODO: Consider using < instead of <=
            if (obj.Position.Y < 0) {
                if (Gravity < 0)
                    f += new Vector(0, obj.Mass * -this.Gravity);
                f += new Vector(0, -obj.Speed.Y * obj.Mass);
            } else if (obj.Position.Y + obj.Diameter > this.Height) {
                if (Gravity > 0)
                    f += new Vector(0, obj.Mass * -this.Gravity);
                f += new Vector(0, -obj.Speed.Y * obj.Mass);
            }

            if (obj.Position.X < 0) {
                // Lineal moment:
                // p = m * v
                // F = dp/dt
                f += new Vector(obj.Mass * -obj.Speed.X, 0);
            } else if (obj.Position.X + obj.Diameter > this.Width) {
                // Lineal moment:
                // p = m * v
                // F = dp/dt
                f += new Vector(obj.Mass * -obj.Speed.X, 0);
            }

            return f;
        }
    }
}

File: cocoa/src/circles/World.swift
//
//  World.swift
//  notifications
//
//  Created by Emilio Cobos Alvarez on 12/4/15.
//  Copyright © 2015 Emilio Cobos Alvarez. All rights reserved.
//

import Foundation

class World {
    var width: CGFloat;
    var height: CGFloat;
    var objects: Array<MovingCircle>;
    var gravity: CGFloat;
    private var precalculatedDeltaFCache: Array<CGVector>?;
    
    init(width: CGFloat, height: CGFloat, gravity: CGFloat, objects: Array<MovingCircle>) {
        self.width = width;
        self.height = height;
        self.gravity = gravity;
        self.objects = objects;
    }
    
    convenience init(width: CGFloat, height: CGFloat, gravity: CGFloat) {
       self.init(width: width, height: height, gravity: gravity, objects: Array())
    }
    
    func tick(var ms: CGFloat) {
        ms /= 1000; // Just seconds
        let len = self.objects.count;
        var accelerationCache = Array<CGVector>(count: len, repeatedValue: CGVectorMake(0.0, 0.0));
        
        self.precalculatedDeltaFCache = Array<CGVector>(count: len, repeatedValue: CGVectorMake(0.0, 0.0));
        
        for i in 0..<len {
            let obj = self.objects[i];
            let force = self.force(obj, myIndex: i, len: len);
            let acceleration = CGVectorMake(force.dx / obj.mass, force.dy / obj.mass);
            
            obj.position.x += ms * (obj.speed.dx + ms * acceleration.dx * 0.5);
            obj.position.y += ms * (obj.speed.dy + ms * acceleration.dy * 0.5);
            
            accelerationCache[i] = acceleration;
            self.calculateSpeedAndPositionFromCollisions(obj, myIndex: i, len: len);
        }
        
        self.precalculatedDeltaFCache = Array<CGVector>(count: len, repeatedValue: CGVectorMake(0.0, 0.0));
        
        for i in 0..<len {
            let obj = self.objects[i];
            let force = self.force(obj, myIndex: i, len: len);
            let newAcceleration = CGVectorMake(force.dx / obj.mass, force.dy / obj.mass);
            
            obj.speed.dx += ms * (accelerationCache[i].dx + newAcceleration.dx);
            obj.speed.dy += ms * (accelerationCache[i].dy + newAcceleration.dy);
        }
    }
    
    func tick(ms: CGFloat, gravity: CGFloat) {
        self.gravity = gravity;
        self.tick(ms);
    }
    
    func force(obj: MovingCircle, myIndex: Int, len: Int) -> CGVector {
        var f = precalculatedDeltaFCache![myIndex];
        f.dy += gravity * obj.mass;
        
        for i in (myIndex + 1)..<len {
            let other = self.objects[i];
            
            let centerToCenter = CGVector(dx: other.position.x - obj.position.x, dy: other.position.y - obj.position.y);
            
            let centerToCenterLenSquared = centerToCenter.dx * centerToCenter.dx + centerToCenter.dy * centerToCenter.dy;
            
            // If they collide
            if centerToCenterLenSquared <= pow(obj.radius + other.radius, 2) {
                f.dx += other.speed.dx * other.mass;
                f.dy += other.speed.dy * other.mass;
                
                f.dx -= obj.speed.dx * obj.mass;
                f.dy -= obj.speed.dy * obj.mass;
                
                precalculatedDeltaFCache![i].dx += obj.speed.dx * obj.mass;
                precalculatedDeltaFCache![i].dy += obj.speed.dy * obj.mass;
                
                precalculatedDeltaFCache![i].dx -= other.speed.dx * other.mass;
                precalculatedDeltaFCache![i].dy -= other.speed.dy * other.mass;
            }
        }
        
        // if object touches boundary
        if obj.position.y - obj.radius < 0 {
            if self.gravity < 0 {
                f.dy -= obj.mass * self.gravity;
            }
            f.dy -= obj.speed.dy * obj.mass;
        } else if obj.position.y + obj.radius > self.height {
            if self.gravity > 0 {
                f.dy -= obj.mass * self.gravity;
            }
            f.dy -= obj.speed.dy * obj.mass;
        }
        
        if obj.position.x - obj.radius < 0 {
            f.dx -= obj.mass * obj.speed.dx;
        } else if obj.position.x + obj.radius > self.width {
            f.dx -= obj.mass * obj.speed.dx;
        }
        return f;
    }
    
    func calculateSpeedAndPositionFromCollisions(obj: MovingCircle, myIndex: Int, len: Int) {
        for i in (myIndex + 1)..<len {
            let other = self.objects[i];
            
            var centerToCenter = CGVector(dx: other.position.x - obj.position.x, dy: other.position.y - obj.position.y);
            
            let centerToCenterLenSquared = centerToCenter.dx * centerToCenter.dx + centerToCenter.dy * centerToCenter.dy;
            
            if centerToCenterLenSquared <= pow(obj.radius + other.radius, 2) {
                let centerToCenterModulus = sqrt(centerToCenterLenSquared);
                
                // Move the objects to the middle of the crash
                // NOTE: This can be heavily optimised, and I expect the compiler to do so
                let a = CGVector(dx: centerToCenter.dx * obj.radius / centerToCenterModulus,
                                 dy: centerToCenter.dy * obj.radius / centerToCenterModulus);
                let b = CGVector(dx: centerToCenter.dx * (centerToCenterModulus - other.radius) / centerToCenterModulus,
                                 dy: centerToCenter.dy * (centerToCenterModulus - other.radius) / centerToCenterModulus);
                let middleX = (a.dx - b.dx) / 2;
                let middleY = (a.dy - b.dy) / 2;
                
                obj.position.x -= middleX;
                obj.position.y -= middleY;
                
                other.position.x += middleX;
                other.position.y += middleY;
                
                // Normalize the vector
                centerToCenter.dx /= centerToCenterModulus;
                centerToCenter.dy /= centerToCenterModulus;
                
                let p = 2 * (obj.speed.dx * centerToCenter.dx + obj.speed.dy * centerToCenter.dy - other.speed.dx * centerToCenter.dx - other.speed.dy * centerToCenter.dy) / (obj.mass + other.mass);
                
                obj.speed.dx = (obj.speed.dx - p * other.mass * centerToCenter.dx) * obj.bounceFactor;
                obj.speed.dy = (obj.speed.dy - p * other.mass * centerToCenter.dy) * obj.bounceFactor;
                
                other.speed.dx = (other.speed.dx + p * obj.mass * centerToCenter.dx) * other.bounceFactor;
                other.speed.dy = (other.speed.dy + p * obj.mass * centerToCenter.dy) * other.bounceFactor;
            }
        }
        // if object touches boundary
        if obj.position.y - obj.radius < 0 {
            obj.speed.dy = abs(obj.speed.dy * obj.bounceFactor);
            obj.position.y = obj.radius;
        } else if obj.position.y + obj.radius > self.height {
            obj.speed.dy = -abs(obj.speed.dy * obj.bounceFactor);
            obj.position.y = self.height - obj.radius;
        }
        
        if obj.position.x - obj.radius < 0 {
            obj.speed.dx = -obj.speed.dx * obj.bounceFactor;
            obj.position.x = obj.radius;
        } else if obj.position.x + obj.radius > self.width {
            obj.speed.dx = -obj.speed.dx * obj.bounceFactor;
            obj.position.x = self.width - obj.radius;
        }
    }
}